libsnark/zk_proof_systems/plonk/plonk.tcc

```diff
@@ -0,0 +1,447 @@
+    // 2**32-th root of unity exists.
+    FieldT omega_base = libff::get_root_of_unity<FieldT>(num_gates);
+#ifdef DEBUG
+    // assert that omega_base is a 2^32-th root of unity in Fq
+    omega_base.print();
+    FieldT temp = libff::power(omega_base, libff::bigint<1>(std::pow(2, 32)));
+    assert(temp == 1);
+#endif // #ifdef DEBUG
+    for (int i = 0; i < (int)num_gates; ++i) {
+        FieldT omega_i = libff::power(omega_base, libff::bigint<1>(i));
+        omega[base][i] = omega_i;
+        FieldT omega_k1_i = omega[base][i] * k1;
+        FieldT omega_k2_i = omega[base][i] * k2;
+        omega[base_k1][i] = omega_k1_i;
+        omega[base_k2][i] = omega_k2_i;
+    }
+}
+
+// copy the roots of unity omega[base], omega[k1] and omega[k2] into
+// a single vector H_gen representing the multiplicative subgroups
+// H, k1 H and k2 H (see [GWC19] Sect. 8); \see
+// plonk_compute_roots_of_unity_omega
+template<typename FieldT>
+void plonk_roots_of_unity_omega_to_subgroup_H(
+    std::vector<std::vector<FieldT>> omega, std::vector<FieldT> &H_gen)
+{
+    assert(omega.size() == NUM_HGEN);
+    assert(omega[0].size() > 0);
+    std::copy(omega[base].begin(), omega[base].end(), back_inserter(H_gen));
+    std::copy(
+        omega[base_k1].begin(), omega[base_k1].end(), back_inserter(H_gen));
+    std::copy(
+        omega[base_k2].begin(), omega[base_k2].end(), back_inserter(H_gen));
+}
+
+// permute the multiplicative subgroup H accordingto the wire
+// permutation: (see [GWC19] Sect. 8), \see
+// plonk_compute_roots_of_unity_omega, \see
+// plonk_roots_of_unity_omega_to_subgroup_H
+template<typename FieldT>
+void plonk_permute_subgroup_H(
+    std::vector<FieldT> &H_gen_permute,
+    std::vector<FieldT> H_gen,
+    std::vector<size_t> wire_permutation)
+{
+    assert(H_gen.size() > 0);
+    for (size_t i = 0; i < H_gen.size(); ++i) {
+        H_gen_permute[i] = H_gen[wire_permutation[i] - 1];
+    }
+}
+// compute the permutation polynomials S_sigma_1, S_sigma_2,
+// S_sigma_2 (see [GWC19], Sect. 8.1)
+template<typename FieldT>
+void plonk_compute_permutation_polynomials(
+    std::vector<polynomial<FieldT>> &S_polys,
+    std::vector<FieldT> H_gen_permute,
+    std::vector<polynomial<FieldT>> L_basis,
+    size_t num_gates)
+{
+    assert(S_polys.size() == NUM_HGEN);
+    assert(S_polys[0].size() == num_gates);
+    assert(H_gen_permute.size() == (NUM_HGEN * num_gates));
+    for (size_t i = 0; i < NUM_HGEN; ++i) {
+        typename std::vector<FieldT>::iterator begin =
+            H_gen_permute.begin() + (i * num_gates);
+        typename std::vector<FieldT>::iterator end =
+            H_gen_permute.begin() + (i * num_gates) + (num_gates);
+        std::vector<FieldT> S_points(begin, end);
+        plonk_interpolate_over_lagrange_basis<FieldT>(
+            S_points, S_polys[i], L_basis);
+    }
+}
+
+// A wrapper for multi_exp_method_BDLO12_signed() for multiplying a
+// single group element in G1 (a point on the curve) to a scalar
```

Comment:
Should this be "multiplying **by** a scalar"?  If so, why use multiexp to do this?  The G1 interface should provide an `operator*` for this.

Reply:
Thanks! Fixed.